compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1444,7 +1444,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // referent for the reference that results is *equal to* the
             // type of the place it is referencing, and not some
             // supertype thereof.
-            let init_ty = self.check_expr_with_needs(init, Needs::maybe_mut_place(m));
+            let mut init_ty = self.check_expr_with_expectation_and_needs(
+                init,
+                ExpectHasType(local_ty),
```

Comment:
Huh, I wonder if this has subtle inference implications. Did you look into this?

Also, why did you add this?

Reply:
This is the call made by `check_expr_coercible_to_type` (which we call in the non-bindings case) - I wanted to align them more closely.